File: internal/app/maintainability.mjs
function analyzeMaintainability(ts, collected) {
  if (process.env.SLICK_MAINTAINABILITY !== "1") return undefined;
  const printer = ts.createPrinter({ removeComments: true, newLine: ts.NewLineKind.LineFeed });

  function metrics(node) {
    const sourceFile = node.getSourceFile();
    const text = node.body?.getText(sourceFile) ?? node.getText(sourceFile);
    const scanner = ts.createScanner(ts.ScriptTarget.Latest, true, ts.LanguageVariant.Standard, text);
    const operators = new Map();
    const operands = new Map();
    let operatorCount = 0;
    let operandCount = 0;
    for (let token = scanner.scan(); token !== ts.SyntaxKind.EndOfFileToken; token = scanner.scan()) {
      const value = scanner.getTokenText();
      if (token === ts.SyntaxKind.Identifier || token === ts.SyntaxKind.PrivateIdentifier ||
          token === ts.SyntaxKind.StringLiteral || token === ts.SyntaxKind.NumericLiteral ||
          token === ts.SyntaxKind.BigIntLiteral || token === ts.SyntaxKind.NoSubstitutionTemplateLiteral ||
          token === ts.SyntaxKind.TrueKeyword || token === ts.SyntaxKind.FalseKeyword ||
          token === ts.SyntaxKind.NullKeyword || token === ts.SyntaxKind.ThisKeyword) {
        operandCount++;
        operands.set(value, (operands.get(value) ?? 0) + 1);
      } else if (![ts.SyntaxKind.OpenBraceToken, ts.SyntaxKind.CloseBraceToken,
          ts.SyntaxKind.OpenParenToken, ts.SyntaxKind.CloseParenToken,
          ts.SyntaxKind.OpenBracketToken, ts.SyntaxKind.CloseBracketToken,
          ts.SyntaxKind.CommaToken, ts.SyntaxKind.SemicolonToken].includes(token)) {
        operatorCount++;
        operators.set(`${token}:${value}`, (operators.get(`${token}:${value}`) ?? 0) + 1);
      }
    }
    const vocabulary = operators.size + operands.size;
    const length = operatorCount + operandCount;
    const volume = vocabulary === 0 || length === 0 ? 0 : length * Math.log2(vocabulary);
    const printed = printer.printNode(ts.EmitHint.Unspecified, node.body ?? node, sourceFile);
    const loc = Math.max(1, printed.split("\n").filter((line) => line.trim() !== "").length);
    return { distinctOperators: operators.size, distinctOperands: operands.size, operatorCount, operandCount, vocabulary, length, volume, loc };
  }

  const results = [];
  for (const functions of collected.functionsByFile.values()) {
    for (const callable of functions) {
      const halstead = metrics(callable.node);
      const safeVolume = Math.max(1, halstead.volume);
      const raw = (171 - 5.2 * Math.log(safeVolume) - 0.23 * callable.complexity - 16.2 * Math.log(halstead.loc)) * 100 / 171;
      results.push({
        symbol: callable.symbol,
        path: callable.path,
        range: callable.range,
        complexity: callable.complexity,
        ...halstead,
        index: Math.max(0, Math.min(100, raw)),
      });
    }
  }
  results.sort((left, right) => {
    const pathOrder = left.path < right.path ? -1 : left.path > right.path ? 1 : 0;
    return pathOrder || left.range.start.offset - right.range.start.offset;
  });
  return { results };
}

File: internal/app/maintainability.go
package app

import "sort"

type MaintainabilityResult struct {
	Symbol            string  `json:"symbol"`
	Path              string  `json:"path"`
	Range             Range   `json:"range"`
	Complexity        int     `json:"complexity"`
	DistinctOperators int     `json:"distinctOperators"`
	DistinctOperands  int     `json:"distinctOperands"`
	OperatorCount     int     `json:"operatorCount"`
	OperandCount      int     `json:"operandCount"`
	Vocabulary        int     `json:"vocabulary"`
	Length            int     `json:"length"`
	Volume            float64 `json:"volume"`
	LOC               int     `json:"loc"`
	Index             float64 `json:"index"`
}

func failingMaintainability(values []MaintainabilityResult, threshold float64) []MaintainabilityResult {
	if threshold <= 0 {
		return nil
	}
	result := []MaintainabilityResult{}
	for _, value := range values {
		if value.Index < threshold {
			result = append(result, value)
		}
	}
	sort.Slice(result, func(i, j int) bool {
		if result[i].Index != result[j].Index {
			return result[i].Index < result[j].Index
		}
		if result[i].Path != result[j].Path {
			return result[i].Path < result[j].Path
		}
		return result[i].Range.Start.Offset < result[j].Range.Start.Offset
	})
	return result
}
